app/Jobs/FetchCollectionActivity.php

```diff
@@ -0,0 +1,125 @@
+use Illuminate\Contracts\Queue\ShouldQueue;
+use Illuminate\Foundation\Bus\Dispatchable;
+use Illuminate\Queue\InteractsWithQueue;
+use Illuminate\Queue\SerializesModels;
+use Illuminate\Support\Facades\DB;
+use Throwable;
+
+class FetchCollectionActivity implements ShouldQueue
+{
+    use Dispatchable, InteractsWithQueue, Queueable, RecoversFromProviderErrors, SerializesModels;
+
+    /**
+     * Create a new job instance.
+     */
+    public function __construct(
+        public Collection $collection,
+        public bool $forced = false,
+    ) {
+        $this->onQueue(Queues::SCHEDULED_NFTS);
+    }
+
+    /**
+     * Execute the job.
+     */
+    public function handle(MnemonicWeb3DataProvider $provider): void
+    {
+        if ($this->collection->isSpam()) {
+            return;
+        }
+
+        if ($this->collection->is_fetching_activity && ! $this->forced) {
+            return;
+        }
+
+        $this->collection->update([
+            'is_fetching_activity' => true,
+        ]);
+
+        $limit = 500;
+
+        $activities = $provider->getCollectionActivity(
+            chain: $this->collection->network->chain(),
+            contractAddress: $this->collection->address,
+            limit: $limit,
+            from: $this->latestActivityTimestamp(),
+        );
+
+        if ($activities->isEmpty()) {
+            $this->collection->update([
+                'is_fetching_activity' => false,
+                'activity_updated_at' => now(),
+            ]);
+
+            return;
+        }
+
+        $formattedActivities = $activities
+            // Sometimes the request is returning transfers that are not labeled as any of the values we expect.
+            // There were times when Mnemonic returned `LABEL_BURN` transfers, so we're filtering them here.
+            ->reject(fn ($activity) => $activity->type === null)
+            ->unique->key()
+            ->map(fn (CollectionActivity $activity) => [
+                'collection_id' => $this->collection->id,
+                'token_number' => $activity->tokenId,
+                'type' => $activity->type->value,
+                'sender' => $activity->sender,
+                'recipient' => $activity->recipient,
+                'tx_hash' => $activity->txHash,
+                'timestamp' => $activity->timestamp,
+                'total_native' => $activity->totalNative,
+                'total_usd' => $activity->totalUsd,
+                'extra_attributes' => json_encode($activity->extraAttributes),
+            ])->toArray();
+
+        DB::transaction(function () use ($formattedActivities, $activities, $limit) {
+            NftActivity::upsert($formattedActivities, uniqueBy: ['tx_hash', 'collection_id', 'token_number', 'type']);
+
+            // If we get the limit it may be that there are more activities to fetch...
+            if ($limit === count($activities)) {
+                self::dispatch($this->collection, forced: true)->afterCommit();
```

Comment:
calling a event inside a db transaction kinda bothers me, doesnt need to create any error but i think does not belongs here, what do you think?

Reply:
I just did that to avoid repetition as we now have simple if-else. It still has `afterCommit` so I don't think it would be a problem. The alternative approach is:

```php
DB::transaction(function () use ($formattedActivities, $activities, $limit) {
    NftActivity::upsert($formattedActivities, uniqueBy: ['tx_hash', 'collection_id', 'token_number', 'type']);

    // If we get the limit it may be that there are more activities to fetch...
    if ($limit !== count($activities)) {
        $this->collection->update([
            'is_fetching_activity' => false,
            'activity_updated_at' => now(),
        ]);
    }
}, attempts: 5);

if ($limit === count($activities)) { // repeating
    self::dispatch($this->collection, forced: true);
}
```

It's still basically the same code, both ways are just fine in my book, what do you think?